CMakeLists.txt

```diff
@@ -27,6 +27,78 @@ ELSE()
     SET(VER_BUILD 0)
 ENDIF()
 
+# Configuration of compilers
+## Compiler flags to add
+### Flags for GCC and Clang
+set(GNU_FLAGS_WARNING_YES "-Wall -Wextra -Wformat=2 -Wpedantic -Wshadow -Wmissing-declarations -Wfloat-equal")
+set(GNU_FLAGS_WARNING_NO  "-Wno-unused-command-line-argument -Wno-documentation -Wno-format-nonliteral")
+set(GNU_FLAGS_WARNING     "${GNU_FLAGS_WARNING_YES} ${GNU_FLAGS_WARNING_NO}")
+set(GNU_FLAGS_ERROR       "-pedantic-errors -Werror") # warning is error
+if (CMAKE_C_COMPILER_ID STREQUAL "GNU" OR
+    CMAKE_C_COMPILER_ID STREQUAL "GCC")
+  set(GNU_FLAGS_SUGGEST   "-Wsuggest-attribute=pure")
+endif()
+set(GNU_FLAGS_DEFAULT     "${GNU_FLAGS_WARNING} ${GNU_FLAGS_ERROR} ${GNU_FLAGS_SUGGEST}")
+if (CMAKE_C_COMPILER_ID STREQUAL "GNU"   AND
+    CMAKE_C_COMPILER_VERSION VERSION_GREATER 4.9)
+  # A color output is easier to read.
+  set(GNU_FLAGS_DEFAULT   "${GNU_FLAGS_DEFAULT} -fdiagnostics-color=auto")
+endif()
+set(GNU_FLAGS_OPTIMIZE    "-O2 -s")
+set(GNU_FLAGS_SECURITY_PI "-fPIC -fPIE -pie")
+if (( CMAKE_C_COMPILER_ID STREQUAL "GNU"   AND
+      CMAKE_C_COMPILER_VERSION VERSION_LESS 4.9 ) OR
+    ( CMAKE_C_COMPILER_ID STREQUAL "Clang" AND
+      CMAKE_C_COMPILER_VERSION VERSION_LESS 3.5 ))
+  set(GNU_FLAGS_SECURITY_STACK "-fstack-protector-all")
+  # Protects a bit "too much", but better that nothing
+else()
+  set(GNU_FLAGS_SECURITY_STACK "-fstack-protector-strong")
+endif()
+set(GNU_FLAGS_SECURITY_COMPILE "-D_FORTIFY_SOURCE=2")
+set(GNU_FLAGS_SECURITY_COMPILE
+  "${GNU_FLAGS_SECURITY_COMPILE} ${GNU_FLAGS_SECURITY_PI}")
+set(GNU_FLAGS_SECURITY_COMPILE
+  "${GNU_FLAGS_SECURITY_COMPILE} ${GNU_FLAGS_SECURITY_STACK}")
+set(GNU_FLAGS_SECURITY_LINK    "-Wl,-z,relro -Wl,-z,now")
+if (WIN32)
+  set(GNU_FLAGS_SECURITY_LINK  "${GNU_FLAGS_SECURITY_LINK} -Wl,dynamicbase") # ASLR
+  set(GNU_FLAGS_SECURITY_LINK  "${GNU_FLAGS_SECURITY_LINK} -Wl,nxcompat")    # DEP
+endif()
+set(GNU_FLAGS_SECURITY  "${GNU_FLAGS_SECURITY_COMPILE} ${GNU_FLAGS_SECURITY_LINK}")
+set(GNU_FLAGS_DEBUG     "-O0 -g -D_FORTIFY_SOURCE=0")
+set(GNU_C_FLAGS_RELEASE "${GNU_FLAGS_OPTIMIZE} ${GNU_FLAGS_SECURITY}")
+set(GNU_C_FLAGS_DEBUG   "${GNU_FLAGS_DEBUG}")
+set(GNU_C_FLAGS         "${GNU_FLAGS_DEFAULT}")
+if (CMAKE_BUILD_TYPE MATCHES "RELEASE" OR
+    CMAKE_BUILD_TYPE MATCHES "release" OR
+    CMAKE_BUILD_TYPE MATCHES "Release")
+  set(GNU_C_FLAGS       "${GNU_C_FLAGS} ${GNU_C_FLAGS_RELEASE}")
+endif()
+### Flags for LLVM/Clang
+set(LLVM_FLAGS_WARNING_YES
+  "-Weverything")
+set(LLVM_FLAGS_WARNING_NO
+  "-Wno-disabled-macro-expansion -Wno-padded -Wno-bad-function-cast -Wno-unreachable-code -Wno-missing-prototypes")
+set(LLVM_FLAGS_WARNING
+  "${LLVM_FLAGS_WARNING_YES} ${LLVM_FLAGS_WARNING_NO}")
+set(LLVM_FLAGS_SANITIZE
+  "-fsanitize=alignment -fsanitize=bool -fsanitize=bounds -fsanitize=enum -fsanitize=float-divide-by-zero -fsanitize=integer-divide-by-zero -fsanitize=null -fsanitize=object-size -fsanitize=shift")
```

Comment:
Many things: First of all, the llvm is just the same than gcc things. It should be enclosed in a better way so that variables get created only if necessary. Secondly, you create sanitize for builds which is a good thing for debug but a bad one for release. And you don't do it for gcc.